// apprunner/lib/apprunner.js
// apprunner.js
// manage code in api folder, handle errors
// © 2012 Harald Rudell <harald.rudell@therudells.com> (http://haraldrudell.com) MIT License

var topapprunner = require('./topapprunner')

var topLevelExport = topapprunner.getOtherAppRunner()
if (topLevelExport) module.exports = topLevelExport
else {
	var anomaly = require('./anomaly')
	var apilist = require('./apilist')
	var appdata = require('./appdata')
	var apperror = require('./apperror')
	var appinit = require('./appinit')
	var requirem = require('./getrequire')
	var rqs = require('./rqs')
	var serverwrapper = require('./serverwrapper')
	var shutdown = require('./appshutdown')

	module.exports = {
		initApp: appinit.initApp,
		getAppData: appdata.getAppData,
		addUriHandler: serverwrapper.addUriHandler,
		addErrorListener: apperror.addErrorListener,
		removeErrorListener: apperror.removeErrorListener,
		shutdown: shutdown.shutdown,
		anomaly: anomaly.anomaly,
		enableAnomalyMail: anomaly.enableAnomalyMail,
		apisReady: apilist.apisReady,
		getApiStates: apilist.getApiStates,
		getRqs: rqs.getRqs,
		getRequire: requirem.getRequire,
		mixIn: requirem.mixIn,
		testIntercept: requirem.testIntercept,
	}
}
