// apprunner/test/test-appinit.js
// test-appinit.js
// © 2012 Harald Rudell <harald.rudell@therudells.com> (http://haraldrudell.com) MIT License

var appinit = require('../lib/appinit')

var appshutdown = require('../lib/appshutdown')
var appdata = require('../lib/appdata')
var apperror = require('../lib/apperror')
var serverwrapper = require('../lib/serverwrapper')
var anomaly = require('../lib/anomaly')
var getrequire = require('../lib/getrequire')
var apionloader = require('../lib/apionloader')
var emailer = require('../lib/emailer')
// http://nodejs.org/api/events.html
var events = require('events')
// http://nodejs.org/api/path.html
var path = require('path')

// https://github.com/haraldrudell/mochawrapper
var assert = require('mochawrapper')

var _log = console.log
var _asi = appshutdown.init
var ae = apperror.addErrorListener
var _ai = anomaly.initAnomaly
var gi = getrequire.init
var dd = apionloader.doOnLoads
var sm = emailer.setSendMail
var ea = anomaly.enableAnomalyMail
var ee = appinit.testEmitter
var iad = appdata.initAppData
var se = serverwrapper.setEmitter

exports['AppInit:'] = {
	'Exports': function () {
		assert.exportsTest(appinit, 2)
	},
	'InitApp': function () {
		var defaults = {
			noInfoLog: true,
			anomaly: false,
		}

		appshutdown.init = function mockAppShutdownInit() {}
		apperror.addErrorListener = function mockAddErrorListener(o) {}
		appdata.initAppData = function mockInitAppData() {}
		emailer.setSendMail = function mockSetSendMail() {}
		serverwrapper.setEmitter = function mockSetEmitter() {}
		getrequire.init = function mockInit() {}
		apionloader.doOnLoads = function mockDoOnLoads() {}

		appinit.initApp(defaults)
	},
	'InitApp InitAnomaly Log': function () {
		var theLog = 3
		var appName = 'APPNAME'
		var defaults = {
			anomaly: {
				noEmail: '1/1/2013',
			},
		}

		appshutdown.init = function mockAppShutdownInit() {}
		apperror.addErrorListener = function mockAddErrorListener(o) {}
		appdata.initAppData = function mockInitAppData() {return {appName: appName, log: theLog}}
		emailer.setSendMail = function mockSetSendMail() {}
		serverwrapper.setEmitter = function mockSetEmitter() {}
		getrequire.init = function mockInit() {}
		apionloader.doOnLoads = function mockDoOnLoads() {}
/*
		var aAddErrorListener = []
		var eAddErrorListener = [appinit.testEmitter()]
		apperror.addErrorListener = function mockAddErrorListener(o) {aAddErrorListener.push(o)}
*/
		function mockSendMail(s, b) {}
		emailer.setSendMail = function mockSetSendMail(x) {assert.equal(x, mockSendMail)}

		var aEnableAnomalyMail = []
		var tzMs = (new Date).getTimezoneOffset() * 60*1e3
		var msPerDay = 24*60*60*1e3
		var dayTimeval = (Math.floor((new Date(defaults.anomaly.noEmail).getTime() - tzMs) / msPerDay) + 1) * msPerDay + tzMs
		var eEnableAnomalyMail = [dayTimeval]
		anomaly.enableAnomalyMail = function mockEnableAnomalyMail(x) {aEnableAnomalyMail.push(x)}

		var aInitAnomaly = []
		var eInitAnomaly = [[{app: appName, noEmail: defaults.anomaly.noEmail}, theLog]]
		anomaly.initAnomaly = function mockInitAnomaly(o, l) {aInitAnomaly.push([o, l])}

		console.log = function () {} // test the console.log statements, too
		appinit.initApp(defaults)
		console.log = _log

//		assert.deepEqual(aAddErrorListener, eAddErrorListener)
		assert.deepEqual(aEnableAnomalyMail, eEnableAnomalyMail)
		assert.deepEqual(aInitAnomaly, eInitAnomaly)
	},
	'after': function () {
		console.log = _log
		appshutdown.init = _asi
		apperror.addErrorListener = ae
		anomaly.initAnomaly = _ai
		getrequire.init = gi
		apionloader.doOnLoads = dd
		emailer.setSendMail = sm
		anomaly.enableAnomalyMail = ea
		appinit.testEmitter(ee)
		appdata.initAppData = iad
		serverwrapper.setEmitter = se
	}
}
