// angular2-universal-preview/dist/client/src/ng_preload_cache.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
/// <reference path="../typings/tsd.d.ts" />
var http_1 = require('angular2/http');
var async_1 = require('angular2/src/core/facade/async');
var lang_1 = require('angular2/src/core/facade/lang');
var angular2_1 = require('angular2/angular2');
exports.PRIME_CACHE = lang_1.CONST_EXPR(new angular2_1.OpaqueToken('primeCache'));
var NgPreloadCacheHttp = (function (_super) {
    __extends(NgPreloadCacheHttp, _super);
    function NgPreloadCacheHttp(_backend, _defaultOptions) {
        _super.call(this, _backend, _defaultOptions);
        this._backend = _backend;
        this._defaultOptions = _defaultOptions;
        this.prime = true;
    }
    NgPreloadCacheHttp.prototype.preload = function (method) {
        var obs = new angular2_1.EventEmitter();
        var newcache = window.ngPreloadCache;
        if (newcache) {
            var preloaded = null;
            try {
                var res;
                preloaded = newcache.shift();
                if (lang_1.isPresent(preloaded)) {
                    var body = preloaded._body;
                    res = new http_1.ResponseOptions(Object.assign({}, preloaded, { body: body }));
                    if (preloaded.headers) {
                        res.headers = new http_1.Headers(preloaded);
                    }
                    preloaded = new http_1.Response(res);
                }
            }
            catch (e) {
                console.log('WAT', e);
            }
            if (preloaded) {
                setTimeout(function () {
                    async_1.ObservableWrapper.callNext(obs, preloaded);
                    // setTimeout(() => {
                    async_1.ObservableWrapper.callReturn(obs);
                    // });
                });
                return obs;
            }
        }
        var request = method();
        // request.observer(obs);
        request.observer({
            next: function (value) {
                async_1.ObservableWrapper.callNext(obs, value);
            },
            throw: function (e) {
                setTimeout(function () {
                    async_1.ObservableWrapper.callThrow(obs, e);
                });
            },
            return: function () {
                setTimeout(function () {
                    async_1.ObservableWrapper.callReturn(obs);
                });
            }
        });
        return obs;
    };
    NgPreloadCacheHttp.prototype.request = function (url, options) {
        var _this = this;
        return this.prime ? this.preload(function () { return _super.prototype.request.call(_this, url, options); }) : _super.prototype.request.call(this, url, options);
    };
    NgPreloadCacheHttp.prototype.get = function (url, options) {
        var _this = this;
        return this.prime ? this.preload(function () { return _super.prototype.get.call(_this, url, options); }) : _super.prototype.get.call(this, url, options);
    };
    NgPreloadCacheHttp.prototype.post = function (url, body, options) {
        var _this = this;
        return this.prime ? this.preload(function () { return _super.prototype.post.call(_this, url, body, options); }) : _super.prototype.post.call(this, url, body, options);
    };
    NgPreloadCacheHttp.prototype.put = function (url, body, options) {
        var _this = this;
        return this.prime ? this.preload(function () { return _super.prototype.put.call(_this, url, body, options); }) : _super.prototype.put.call(this, url, body, options);
    };
    NgPreloadCacheHttp.prototype.delete = function (url, options) {
        var _this = this;
        return this.prime ? this.preload(function () { return _super.prototype.delete.call(_this, url, options); }) : _super.prototype.delete.call(this, url, options);
    };
    NgPreloadCacheHttp.prototype.patch = function (url, body, options) {
        var _this = this;
        return this.prime ? this.preload(function () { return _super.prototype.patch.call(_this, url, body, options); }) : _super.prototype.patch.call(this, url, body, options);
    };
    NgPreloadCacheHttp.prototype.head = function (url, options) {
        var _this = this;
        return this.prime ? this.preload(function () { return _super.prototype.head.call(_this, url, options); }) : _super.prototype.head.call(this, url, options);
    };
    NgPreloadCacheHttp = __decorate([
        angular2_1.Injectable(), 
        __metadata('design:paramtypes', [http_1.ConnectionBackend, http_1.RequestOptions])
    ], NgPreloadCacheHttp);
    return NgPreloadCacheHttp;
})(http_1.Http);
exports.NgPreloadCacheHttp = NgPreloadCacheHttp;
exports.NG_PRELOAD_CACHE_PROVIDERS = [
    angular2_1.provide(http_1.Http, {
        useFactory: function (xhrBackend, requestOptions) {
            return new NgPreloadCacheHttp(xhrBackend, requestOptions);
        },
        deps: [http_1.XHRBackend, http_1.RequestOptions]
    })
];
